Fix build errors from PR#1 and stop calling completion repeatedly

shepazon opened the pull request:
Updated the code to avoid calling the completion handler more than once when getting the pool ID, and fixed some build errors  from the previous PR.

shepazon committed:
Reworked further to fix problems from Nicki's PR and to not call completion more than once. But this is using some state variables that I want to eradicate. Committing to preserve this code in case the changes I'm about to try backfire.

shepazon committed:
Finished revising to avoid completion handler being called more than once for a given path through the getIdentityPoolID() function. Hopefully this code doesn't violate any sensibilities. :)

kneekey23 commented on Tests/CognitoIdentityDemoTests/CognitoIdentityDemoTests.swift:
```diff
@@ -19,18 +19,26 @@ final class CognitoIdentityDemoTests: XCTestCase {
-        let firstPoolID = identityTester!.getIdentityPoolID(name: "testCreateThenFind", createIfMissing: true)
-        XCTAssertNotNil(firstPoolID, "Unable to create or obtain test pool")
+        var firstPoolID: String?
+        var secondPoolID: String?
```
no need for these. use XCTExpectation to fulfill async expectations in tests. :)

shepazon replied:
How can you compare the result of two tests without storing the value of at least the first result until the second result arrives?

kneekey23 replied:
the calls should be wrapped in each other, welcome to callback hell. you cant really call the second one until the first has completed. and since you wont know if its completed unless you make the call to the second inside the completion handler of the first.
```
        identityTester?.getIdentityPoolID(name: "testCreateThenFind", createIfMissing: true) { poolID in
            let firstPoolId = poolID
            XCTAssertNotNil(firstPoolId, "Unable to create or obtain test pool")
              identityTester?.getIdentityPoolID(name: "testCreateThenFind", createIfMissing: false) { poolID in
                  let secondPoolID = poolID
                  XCTAssertNotNil(secondPoolID, "Unable to find test pool")
                  XCTAssertTrue(firstPoolID == secondPoolID, "Found pool ID doesn't match created pool ID")
              }
        }


```

shepazon committed:
Implement changes based on code from wooj

wooj2 commented on Sources/CognitoIdentityDemo/CognitoIdentityDemo.swift:
```diff
@@ -47,66 +54,69 @@ class CognitoIdentityDemo {
-                                completion(poolId)
-                                break
-                            }
+                        let foundID = identityPools.filter {
+                            $0.identityPoolName == name
+                        }
+                        .map {
+                            $0.identityPoolId
+                        }
+                        .first ?? nil
+                        
+                        if let foundID = foundID {
+                            poolID = foundID
```
This can be written more succinctly as:
```
                case .success(let output):
                    poolID = output.identityPools?
                        .filter { $0.identityPoolName == name }
                        .map { $0.identityPoolId }
                        .first ?? nil
                    token = output.nextToken
```

shepazon replied:
I originally wrote something similar but it didn't work right, so eventually I gave up and did it the way shown above. I'm applying your suggestion now.

wooj2 commented on Sources/CognitoIdentityDemo/CognitoIdentityDemo.swift:
```diff
@@ -47,66 +54,69 @@ class CognitoIdentityDemo {
+        self.getIdentityPoolID(name: name) { foundID in
+            if let foundID = foundID {
+                completion(foundID)
+            } else if createIfMissing {
+                self.createIdentityPool(name: name, completion: completion)
```
This code feels awkward, because it's possible for us to fail the call to `listIdentityPools`, and then attempt to create IdentityPool.

shepazon replied:
That's actually the point, though. To create the pool if it doesn't exist. Though you're right that there's no way to differentiate errors from "not found" here. Which is a good reason why the code ought to be throwing an exception rather than returning nil for errors. But I hesitate to go through another round of changes to the code right at this moment, so instead I would prefer to note the problem and come back and correct this post re:Invent (unless I find I have spare time beforehand to try to get it done).

wooj2 commented on Sources/CognitoIdentityDemo/CognitoIdentityDemo.swift:
```diff
@@ -47,66 +54,69 @@ class CognitoIdentityDemo {
     func createIdentityPool(name: String, completion: @escaping (String?) -> Void) {
-        
         let cognitoInputCall = CreateIdentityPoolInput(
             developerProviderName: "com.exampleco.CognitoIdentityDemo",
```
Seems like the `developerProviderName` should be something injectable to this class, rather than hardcoded in this function.

shepazon replied:
Yeah, that was intentional here though I see your point. Unless there's a strong objection I will revisit this after re:Invent.

wooj2 commented on Sources/CognitoIdentityDemo/CognitoIdentityDemo.swift:
```diff
@@ -47,66 +54,69 @@ class CognitoIdentityDemo {
             case .success(let output):
-                if let poolId = output.identityPoolId {
-                    completion(poolId)
+                if let poolID = output.identityPoolId {
+                    completion(poolID)
                 }
```
Looks like the completion handler doesn't get called if we succeed calling createIdentityPool, but don't get back an identityPoolId.

Rather than have a callback of `(String?) -> Void`, we can use a `Result` type:

shepazon replied:
I've spent some time (more than I should have) trying to get this to work and while I know it can, I don't think I have time to deal with this anymore. Tomorrow I'm going to have to either remove all the code snippets or leave them as is until after re:Invent and finish fixing them then. Let me know how egregious this is.

wooj2 commented on Sources/CognitoIdentityDemo/CognitoIdentityDemo.swift:
```diff
@@ -21,15 +22,21 @@ class CognitoIdentityDemo {
     }
     
-    /// Returns the ID of the identity pool with the specified name.
+    /// Returns the ID of the identity pool with the specified name by calling the completion handler with
```
Please remove the call to:
```
        SDKLoggingSystem.add(logHandlerFactory: CognitoIdentityClientLogHandlerFactory(logLevel: .info))
```
from the `init` function.  This sort of call should be left up to the developer who is instantiating `CognitoIdentityDemo`

wooj2 commented on Sources/CognitoIdentityDemo/CognitoIdentityDemo.swift:
```diff
@@ -47,66 +54,69 @@ class CognitoIdentityDemo {
-                completion(nil)
             case .failure(let error):
-                completion(nil)
                 print("ERROR: ", dump(error, name: "Error attempting to create the identity pool"))
```
Is there a strong reason to use `print` over swift-log?  we can easily do something like:
```
import Logging

...
    let logger: Logger = Logger(label: "CognitoIdentityDemo")
```
and then:
```
logger.error("....")
```

shepazon replied:
The main reason I'm using print is that I figured doing so separates my output from the logging, since my code is supposed to be a stand-in for some kind of functional code rather than a true log entry.

shepazon committed:
Addresses some of @wooj2's suggestions
Removes the logging setup line left over from debugging. Reworked the demo methods to throw instead of returning nil for errors; this fixes the uncertainty as to whether nil means "error" or "not found."

shepazon commented:
I've fixed what I can with the time I have left. The rest will have to wait until next week, unfortunately.
